src/sdformat_urdf.cpp

```diff
@@ -0,0 +1,466 @@
+      auto urdf_material = std::make_shared<urdf::Material>();
+      // sdf materials don't have names, so assign it the visual's name and hope that's unique enough
+      urdf_material->name = sdf_visual->Name();
+      // Color support is pretty limited in urdf, just take the ambient (color when no light applied)
+      urdf_material->color.r = sdf_material->Ambient().R();
+      urdf_material->color.g = sdf_material->Ambient().G();
+      urdf_material->color.b = sdf_material->Ambient().B();
+      urdf_material->color.a = sdf_material->Ambient().A();
+
+      urdf_visual->material = urdf_material;
+    }
+
+    if (0u == vi) {
+      urdf_link->visual = urdf_visual;
+    }
+    urdf_link->visual_array.push_back(urdf_visual);
+  }
+
+  for (uint64_t vi = 0; vi < sdf_link.CollisionCount(); ++vi) {
+    const sdf::Collision * sdf_collision = sdf_link.CollisionByIndex(vi);
+    if (!sdf_collision) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get collision on lilnk [" + sdf_link.Name() + "]");
+      return nullptr;
+    }
+
+    auto urdf_collision = std::make_shared<urdf::Collision>();
+
+    urdf_collision->name = sdf_collision->Name();
+
+    // URDF collision is relative to link origin
+    ignition::math::Pose3d collision_pose;
+    sdf::Errors pose_errors = sdf_collision->SemanticPose().Resolve(collision_pose, sdf_link.Name());
+    if (!pose_errors.empty()) {
+      errors.insert(errors.end(), pose_errors.begin(), pose_errors.end());
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to get transfrom from collision [" + sdf_collision->Name()
+        + "] to link [" + sdf_link.Name() + "]");
+      return nullptr;
+    }
+    urdf_collision->origin = convert_pose(collision_pose);
+
+    urdf_collision->geometry = convert_geometry(*sdf_collision->Geom(), errors);
+    if (!urdf_collision->geometry) {
+      errors.emplace_back(
+        sdf::ErrorCode::STRING_READ,
+        "Failed to convert geometry on collision [" + sdf_collision->Name()+ "]");
+      return nullptr;
+    }
+
+    if (0u == vi) {
+      urdf_link->collision = urdf_collision;
+    }
+    urdf_link->collision_array.push_back(urdf_collision);
+  }
+
+  return urdf_link;
+}
+
+urdf::JointSharedPtr
+sdformat_urdf::convert_joint(const sdf::Joint & sdf_joint, sdf::Errors & errors)
+{
+  urdf::JointSharedPtr urdf_joint = std::make_shared<urdf::Joint>();
+
+  urdf_joint->name = sdf_joint.Name();
+
+  switch (sdf_joint.Type()) {
+    case sdf::JointType::CONTINUOUS:
+      urdf_joint->type = urdf::Joint::CONTINUOUS;
+      break;
+    case sdf::JointType::REVOLUTE:
+      urdf_joint->type = urdf::Joint::REVOLUTE;
+      break;
+    case sdf::JointType::FIXED:
+      urdf_joint->type = urdf::Joint::FIXED;
+      break;
+    case sdf::JointType::PRISMATIC:
+      urdf_joint->type = urdf::Joint::PRISMATIC;
+      break;
+    case sdf::JointType::INVALID:
```

Comment:
Dumb question: Why even specify these cases as fall-through when they all feed into the `default` case?

It seems more harmful to have them here, as someone may naively fill in an intermediate case (e.g. `GEARBOX`) and mess up the fall-through?

Reply:
It's to show they're known and intentionally handled by the default case instead of accidentally handled by the default case.

I hope a contribution adding code in the middle would be caught in code review, and maybe caught with unit testing.

Reply:
Changed fall through comments to make it easier to see the cases are all fall through to `default`

ecb6e32e8ede94cf00ac7051510312acc1f33af5